### DISCUSSION TITLE
Real-world open-source project based on .NET 6 with DDD, ES, CQRS and Testing concepts

### DISCUSSION THREAD

**Discussion by hamed-shirbandi** [techexcellenceio/techexcellence#11] [General]
Hello to the community.

I would like to introduce a free and open-source project named [**TaskoMask** ](https://github.com/hamed-shirbandi/TaskoMask) that I have been working on it for about 2 years.
It is based on .NET 6 with Microservices, DDD, CQRS, Event Sourcing and Testing Concepts.
There are some [documentations ](https://github.com/hamed-shirbandi/TaskoMask#documentation) you can use to learn more about it. Also, the project is [online](http://taskomask.ir) and you can see how the things work.

![taskomask](https://github.com/hamed-shirbandi/TaskoMask/raw/master/docs/images/Shots/taskomask-all-in-one-mobile.jpg)

# Design

  * ### Onion Architecture
    <details>
      <summary>click for details</summary>

    - Architecture was Onion until this [last commit](https://github.com/hamed-shirbandi/TaskoMask/tree/af7f7418c7811ecf2db3bb9324bd070e37eb7a82)
    </details>
  * ### Clean Architecture
  * ### Vertical Slice Architecture
  * ### Feature Folder Structure
  * ### Monolithic Architecture
    <details>
      <summary>click for details</summary>

    - Architecture was Monolithic until this [last commit](https://github.com/hamed-shirbandi/TaskoMask/tree/af7f7418c7811ecf2db3bb9324bd070e37eb7a82)
    </details>
  * ### Microservices Architecture
    <details>
      <summary>click for details</summary>

    - [Strangler application pattern](https://microservices.io/refactoring/)
    - [Decompose by subdomain](https://microservices.io/patterns/decomposition/decompose-by-subdomain.html)
    - [Database per service](https://microservices.io/patterns/data/database-per-service.html)
    - [Saga](https://microservices.io/patterns/data/saga.html)
    - [API Composition](https://microservices.io/patterns/data/api-composition.html)
    - [Docker](https://www.docker.com/)
    - [Docker-Compose](https://docs.docker.com/compose/)
    - [Kubernetes](https://kubernetes.io/)
    - [Messaging](https://microservices.io/patterns/communication-style/messaging.html) : MassTransit (RabbitMQ)
    - [Remote Procedure Call](https://microservices.io/patterns/communication-style/rpi.html) : Grpc.AspNetCore
    - [Idempotent Consumer](https://microservices.io/patterns/communication-style/idempotent-consumer.html)
    - [API Gateway](https://microservices.io/patterns/apigateway.html) : Ocelot
    - [Backend for front-end](https://microservices.io/patterns/apigateway.html)
    - [Service discovery](https://microservices.io/patterns/3rd-party-registration.html) : Kubernetes - Consul
    - [Circuit Breaker](https://microservices.io/patterns/reliability/circuit-breaker.html) : Polly
    - [Log aggregation](https://microservices.io/patterns/observability/application-logging.html) : Serilog - Seq
    - [Application metrics](https://microservices.io/patterns/observability/application-metrics.html) : Opentelemetry-dotnet - Prometheus
    - [Distributed tracing](https://microservices.io/patterns/observability/distributed-tracing.html) : Opentelemetry-dotnet - Jaeger
    - [Health check API](https://microservices.io/patterns/observability/health-check-api.html) : AspNetCore.HealthChecks
    - [IDP](https://en.wikipedia.org/wiki/Identity_provider) : DuendeSoftware IdentityServer
    </details>
  * ### Testing (TDD & BDD)
    <details>
      <summary>click for details</summary>

    - [Unit Testing](https://github.com/hamed-shirbandi/TaskoMask/tree/master/Src/Tests/Unit) : xUnit, FluenAssertion, NSubstitute
    - [Integration Testing](https://github.com/hamed-shirbandi/TaskoMask/tree/master/Src/Tests/Integration)
    - [API Testing](https://github.com/hamed-shirbandi/TaskoMask/tree/master/Src/Tests/Acceptance/Tests.Acceptance.API)
    - [UI Testing](https://github.com/hamed-shirbandi/TaskoMask/tree/master/Src/Tests/Acceptance/Tests.Acceptance.UI): Selenium
    - [Acceptance Testing](https://github.com/hamed-shirbandi/TaskoMask/tree/master/Src/Tests/Acceptance) : Gherkin, SpecFlow
    - [Screenplay Pattern](https://serenity-js.org/handbook/design/screenplay-pattern.html#:~:text=The%20Screenplay%20Pattern%20is%20a,testing%20and%20software%20engineering%20habits.) : Suzianna
    - Well written tests organized in :
        - [Business Rule Layer](https://www.oreilly.com/library/view/bdd-in-action/9781617291654/)
        - [Business Flow Layer](https://www.oreilly.com/library/view/bdd-in-action/9781617291654/)
        - [Technical Layer](https://www.oreilly.com/library/view/bdd-in-action/9781617291654/)
    - [Object Mother Pattern](http://xunitpatterns.com/Test%20Helper.html#Object%20Mother)
    - Test Data Builder
    - Test Hooks
    - Test Doubles
    - Dummy
    - Stub
    - Mock
    - Teardown
        - [Sandbox](http://xunitpatterns.com/Database%20Sandbox.html)
    - Fixture Management
        - [Fresh](http://xunitpatterns.com/Fresh%20Fixture.html)
        - [Shared](http://xunitpatterns.com/Shared%20Fixture.html)
        - [Transient](http://xunitpatterns.com/Fresh%20Fixture.html#Transient%20Fresh%20Fixture)
        - [Persistent](http://xunitpatterns.com/Persistent%20Fixture%20Management.html)
    - Verification
        - [State Verification](http://xunitpatterns.com/ResultVerification.html)
        - [Output/Value Verification](http://xunitpatterns.com/ResultVerification.html)
        - [Interaction/Behavior Verification](http://xunitpatterns.com/ResultVerification.html)
    </details>
  * ### DDD
    <details>
      <summary>click for details</summary>

    - Rich Domain Model (for core domain)
    - Anemic Domain Model (for less important subdomains)
    - Aggregate
    - Value Object
    - Domain Event
    - Domain Service
    - Always Valid Domain Model
    - Invariants
    - Specification
    - Factory Method
    - Optimistic Concurrency
    - Separate Domain Model and Data Model
    </details>
  * ### CQRS
    <details>
      <summary>click for details</summary>

    - Separate Read and Write Model
    - Separate Read Side DB and Write Side DB
    </details>
  * ### Use Case Driven Development
  * ### Event Sourcing
  * ### Unit of Work
  * ### Repository
  * ### Notification

# Implementation
  * ### Back-end:
    <details>
      <summary>click for details</summary>

      - .Net 6 
      - C#
      - ASP.NET Web API
      - ASP.NET MVC
      - ASP.NET Identity
      -	MongoDB
      -	Redis
      - [Ocelot](https://ocelot.readthedocs.io/) : .NET core API Gateway
      - [DuendeSoftware IdentityServer](https://docs.duendesoftware.com/identityserver/v6) : OpenID Connect and OAuth 2.x framework for ASP.NET Core
      - [MassTransit](https://masstransit-project.com/) : a framework on top of message transports such as RabbitMQ 
      - [xUnit](https://xunit.net/) : testing framework
      -	[FluenAssertion](https://fluentassertions.com/) : write fluent assertions
      - [NSubstitute](https://nsubstitute.github.io/) : to make test double (Mock, stub, fake, spy)
      - [Gherkin](https://specflow.org/learn/gherkin/) : use native language to describe test cases
      - [SpecFlow](https://www.nuget.org/packages/SpecFlow.xUnit/) : turns Gherkin scenarios into automated tests
      - [Suzianna](https://github.com/suzianna/Suzianna) : writing acceptance tests, using Screenplay Pattern
      - [Selenium](https://www.nuget.org/packages/Selenium.WebDriver/) : supporting browser automation
      -	[MediatR](https://github.com/jbogard/MediatR) : simple mediator implementation
      -	[Grpc.AspNetCore](https://www.nuget.org/packages/Grpc.AspNetCore/) : gRPC library for ASP.NET Core
      -	[AutoMapper](https://automapper.org/) : an object-object mapper
      -	[FluentValidation](https://docs.fluentvalidation.net/en/latest/) : building strongly-typed validation rules
      -	[Swagger](https://www.nuget.org/packages/Swashbuckle.AspNetCore) : expose Swagger JSON endpoints from APIs
      -	[Serilog](https://serilog.net/) : provides diagnostic logging
      - [AspNetCore.HealthChecks](https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks) : ASP.NET Core Health Check
      -	[MvcPagedList.Core](https://www.nuget.org/packages/MvcPagedList.Core/) : easily paging in ASP.NET Core MVC
      -	[EasyCaching](https://github.com/dotnetcore/EasyCaching) : caching library
    </details>
  * ### Front-end:
    <details>
      <summary>click for details</summary>

      - Blazor
        - Blazor Server ([last commit](https://github.com/hamed-shirbandi/TaskoMask/tree/a6f036f91c2185861209191d9bb3e4ae01665f46/Src/Presentation/3-UI/UserPanel))
            - Cookie Authentication without ASP.NET Identity
        - Blazor WebAssembly (standalone)
            - JWT Authentication
        - Comunication between components
        - Local Storage
        - Consume REST API
        - Retry using HttpClientRetryHelper
        - Handle Drag and Drop
        - Using Modal, Toast, etc.
      -	.HTML
      -	CSS
      -	JavaScript 
      -	JQuery
      -	Bootstrap
      -	Jquery.noty
      -	Chart.js
    </details>
  * ### Some other Features:
    <details>
      <summary>click for details</summary>

      -	Caching Behavior using Pipeline Pattern
      -	Validation Behavior using Pipeline Pattern (Check both Fluent Validation and Data Annotation Validation)
      -	Event Storing Behavior using Pipeline Pattern
      - Exception Handling
      -	Cookie Authentication
      -	JWT Authentication
      -	Role Permission Base User Management without ASP.NET Identity
      -	Swagger UI with JWT Support
    </details>

I would appreciate any support, contribute, help and advice ❤️🙏

**Comment by valentinajemuovic** [comment]
Thanks @hamed-shirbandi for your iniative and for making this GitHub project.

I have some questions regarding Libraries & Presentation:

**1-Infrastructure**

**Infrastructure.CrossCutting**
- What are the pros and cons of having interface IInMemoryBus vs just IBus?
- How did you choose to use IMappingAction compared to using just Profile? (I ask this since for me Profile was enough in the past)

**Infrastructure.Data**
- How did you decide to make the data generators here, is this just some sample data used for demos?
- What's the purpose of IMongoDbContext interface what is it abstracting?
- How did you decide to have IBaseRepository versus having separate like IActivityRepository? I see the expectation that then all the repositories will have the same CRUD operations.

**3-Domain**

**Domain**
- How did you decide to make public setters on the domain entities? Also what if someone sets invalid values?
- I noticed base.SetId(id), what about the possiblity of calling the base constructor?
- What is the purpose of SetAsDeleted(), SetAsRecycled(), SetAsUpdated()?

**Domain.Core**
- How did you decide to put JSonIgnore in the DomainEvent, compared to putting it into infrastructure?
- In the code there is DateTime.Now in the DomainEvent, so this would make the test non-deterministic, how do you handle this? (I ask this since I stopped using DateTime.Now in code directly, instead I have a IDateTimeService)

**Presentation**

**UserPanelAPI**
- BoardController - I see BoardInsertDto used for both Create and Update, do you always do it that way?
- In the route for HttpPut for BoardController.Update, would you use route parameters for the id?

**Reply by hamed-shirbandi** [reply]
Thanks @valentinacupac for the time that you spent on my repository. I really appreciate it.
I got so happy for your feedbacks and I will consider all your advices.

Here is some explanation about the questions:

#### [What are the pros and cons of having interface IInMemoryBus vs just IBus?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
I just like the contributors to know there is just an in-memory bus to sending messages when they see the IInMemoryBus is injected as a dependency. If I add new one, then it is better to use IBus as you said.

#### [How did you choose to use IMappingAction compared to using just Profile?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
MappingAction is used when you need to inject some dependencies like IConfiguration to be used in mapping. Profile doesn't support it.

#### [How did you decide to make the data generators here, is this just some sample data used for demos?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
Yes, just for a demo.

#### [What's the purpose of IMongoDbContext interface what is it abstracting?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
It abstracts some needed MongoDB functionalities and hides implementation detail and it is used to share a unique instance of DbContext during each request scope.

#### [How did you decide to have IBaseRepository versus having separate like IActivityRepository?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
IBaseRepository provides the same functionalities for all entities, but each entity (or aggregate) can have some specific functionalities. For example getting a user by its userName instead of its id in IUserRepository.

#### [How did you decide to make public setters on the domain entities?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
As I have described on [documents ](https://github.com/hamed-shirbandi/TaskoMask/tree/master/Src/Libraries/3-Domain/Domain) related to Domain, I have tried to cover some examples around Rich Domain and Anemic Domain Modeling for implementing Generic and Supporting and Core Domains. I have used Anemic Models to implement less important subdomains (like Authorization and Membership) and Rich Models for the core domain (Workspace).

#### [I noticed base.SetId(id), what about the possiblity of calling the base constructor?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
base.SetId(id) is used for the entities that need to define a special id. Also it is protected and can't be called from the outside of the entities. Calling the base constructor will generate a random id for the entity. So each entity can decide it needs to set a special id or not.

#### [What is the purpose of SetAsDeleted(), SetAsRecycled(), SetAsUpdated()?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
They are just some behaviors related to domain.

#### [How did you decide to put JsonIgnore in the DomainEvent, compared to putting it into infrastructure?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
I just needed to skip some properties when doing Serialization. I could do it by implementing a helper class in infrastructure or by using JsonIgnore attribute in the main model. I chose the simpler. If I see a problem during the tests, I will refactor it.

#### [In the code there is DateTime.Now in the DomainEvent, so this would make the test non-deterministic, how do you handle this?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
I have been following test-last approach in the project and I will refactor it by using IDateTimeService during the tests just like what you adviced.

#### [BoardController - I see BoardInsertDto used for both Create and Update, do you always do it that way?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
I have tried to prepare a quick UI client and that is why most of the forms and input models are the same. I will continue developing the clients and then will separate the models during implementing the real features.

#### [In the route for HttpPut for BoardController.Update, would you use route parameters for the id?](https://github.com/valentinacupac/community/discussions/11#discussioncomment-2766478)
I will refactor it and follow the other REST conventions when I focus on clients.
